### DISCUSSION TITLE
Issues with serializing BitArray

### DISCUSSION THREAD

**Discussion by Jake9768** [thepirat000/Audit.NET#616] [Q&A]
For us, running an AuditScope on our EF models runs fine on an entity that doesn't have a BitArray in it, but throws an exception on the entity that does have a BitArray that is attempted to be serialized into an audit event.

Exception we are getting after running an audit scope through our entity

`The collection type 'System.Collections.BitArray is abstract, an interface, or is read only, and could not be instantiated and populated. Path: $.PrivilegeSet | LineNumber: 0 | BytePositionInLine: 17."`

Our entity setup for PrivilegeSet

`
public Role ( BitArray privilegeSet )
{
    this.privilegeSet = privilegeSet;
}

private BitArray privilegeSet;
public BitArray PrivilegeSet
{
   get => privilegeSet;
   set => Set(ref this.privilegeSet, value, RoleField.PrivilegeSet);
}`

We tried serializing a basic BitArray using both Json.NET and NewtonSoft.JSON and both worked without issue. Our project using Audit.NET is targeting NET 7.0. Any ideas on why we are getting this error with the BitArray property here when running it through an audit scope?

Thank you.

**Comment by thepirat000** [comment]
Can you share the stack trace and your data provider configuration?

**Comment by Jake9768** [comment]
Using the Cosmos Data Provider.

> Configuration.Setup()
>             .UseAzureCosmos(config => config
>             .Endpoint(CosmosEndpoint)
>             .AuthKey(CosmosAuthKey)
>             .Database(CosmosDbName)
>             .Container(CosmosContainer)
>             .ClientOptions(options => options.ConnectionMode = Microsoft.Azure.Cosmos.ConnectionMode.Gateway));

>    at System.Text.Json.ThrowHelper.ThrowNotSupportedException(ReadStack& state, Utf8JsonReader& reader, NotSupportedException ex) in /_/src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs:line 509
>    at System.Text.Json.Serialization.Converters.IEnumerableConverter`1.CreateCollection(Utf8JsonReader& reader, ReadStack& state, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverter.cs:line 33
>    at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonCollectionConverter.cs:line 80
>    at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:line 276
>    at System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`5.TryRead[TArg](ReadStack& state, Utf8JsonReader& reader, JsonParameterInfo jsonParameterInfo, TArg& arg) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:line 65
>    at System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`5.ReadAndCacheConstructorArgument(ReadStack& state, Utf8JsonReader& reader, JsonParameterInfo jsonParameterInfo) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs:line 37
>    at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:line 49
>    at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:line 276
>    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs:line 60
>    at System.Text.Json.Serialization.JsonConverter`1.ReadCoreAsObject(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs:line 13
>    at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo jsonTypeInfo, Nullable`1 actualByteCount) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Helpers.cs:line 49
>    at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo jsonTypeInfo) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 387
>    at System.Text.Json.JsonSerializer.Deserialize(String json, Type returnType, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 133
>    at Audit.Core.JsonAdapter.Deserialize(String json, Type type)
>    at Audit.Core.AuditDataProvider.Serialize[T](T value)
>    at Audit.Core.AuditScope..ctor(AuditScopeOptions options)
>    at Audit.Core.AuditScope.Create(String eventType, Func`1 target, Object extraFields)


Apologize for the formatting, doesnt seem to maintain the new lines on a quote/codeblock. Also just doing a 

var test = JsonSerializer.Serialize(entity);

works just fine for our bitArray.

**Comment by thepirat000** [comment]
The problem is that System.Text.Json cannot properly deserialize BitArray, please check [this SO answer](https://stackoverflow.com/a/58279913/122195).

One option is to use [Audit.NET's Newtonsoft.Json adapter](https://github.com/thepirat000/Audit.NET/blob/master/README.md#alternative-serialization-mechanism) with the custom JsonConverter from that answer:

```
Install-Package Audit.NET.JsonNewtonsoftAdapter
```

```c#
Audit.Core.Configuration.Setup()
    .JsonNewtonsoftAdapter(new JsonSerializerSettings()
    {
        Converters = new List<JsonConverter>() { new BitArrayConverter() }
    });
```

**Reply by Jake9768** [reply]
This got it working with the custom BitArrayConverter() from the SO thread. Thanks.
We did try following the documentation awhile back to switch to Newtonsoft instead of System.Text.Json, but without the custom converter, that didn't work either oddly enough. Still a bit uncertain why the custom converter is needed as both System.Text.Json and Newtonsoft both serialize our entity fine standalone, but no worries as it's working now. Thanks again for the assistance.
